### PR TITLE
Debug

### PR DESCRIPTION
**Pull request by matklad**

### CODE REVIEW

**Comment by matklad** [comment]
@flodiebold what do you think about this setup?

File: crates/ra_hir/src/debug.rs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+//! * when debugging, be aware that interference is possible.
+
+use std::{cell::Cell, fmt};
+
+use ra_db::{CrateId, FileId};
+
+use crate::{db::HirDatabase, Crate, Module, Name};
+
+impl Crate {
+    pub fn debug(self, db: &impl HirDebugDatabase) -> impl fmt::Debug + '_ {
```
Comment (flodiebold) [inline crates/ra_hir/src/debug.rs]: why not start out making this a trait? :thinking:

File: crates/ra_hir/src/debug.rs [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        let file_id = module.definition_source(self).file_id.original_file(self);
+        let path = self.file_path(file_id);
+        fmt.debug_struct("Module")
+            .field("name", &module.name(self).unwrap_or_else(Name::missing))
+            .field("path", &path.unwrap_or_else(|| "N/A".to_string()))
+            .finish()
+    }
+}
+
+fn debug_fn(f: impl FnOnce(&mut fmt::Formatter<'_>) -> fmt::Result) -> impl fmt::Debug {
```
Comment (flodiebold) [inline crates/ra_hir/src/debug.rs]: couldn't this be just a `Fn(&mut fmt::Formatter)`, then the `Cell` dance wouldn't be necessary?

File: crates/ra_ide_api/src/db.rs [hunk 3]
```diff
@@ -90,3 +106,16 @@ fn line_index(db: &impl ra_db::SourceDatabase, file_id: FileId) -> Arc<LineIndex
     let text = db.file_text(file_id);
     Arc::new(LineIndex::new(&*text))
 }
+
+#[derive(Debug, Default, Clone)]
+pub(crate) struct DebugData {
+    pub(crate) root_paths: FxHashMap<SourceRootId, String>,
+    pub(crate) crate_names: FxHashMap<CrateId, String>,
```
Comment (flodiebold) [inline crates/ra_ide_api/src/db.rs]: wouldn't we want to have access to these things in some non-debugging cases as well?

File: crates/ra_hir/src/debug.rs (hunk 1 above)
Comment (matklad) [reply crates/ra_hir/src/debug.rs]: inherent method is more convenient for debugging as you don't have to add an import, and we don't have to generalize over this functionality yet

File: crates/ra_hir/src/debug.rs (hunk 2 above)
Comment (matklad) [reply crates/ra_hir/src/debug.rs]: heh, of course it could, good catch!

File: crates/ra_ide_api/src/db.rs (hunk 3 above)
Comment (matklad) [reply crates/ra_ide_api/src/db.rs]: I think only for debuginfo generation (which is far off) and for error reporting (which happens outside of hir). So yeah, I think we should maybe name this thing differently, but it shouldn't be accessible to HIR in the same way the source database is accessible .

**Comment by matklad** [comment]
>Would it maybe be possible to do this with 'untracked' reads (not sure if Salsa supports that), so you can't accidentally modify the dep graph so much?

Yeah, we should add something like this, but salsa doesn't support this at the moment, and it seems not too trivial. 

bors r+

File: crates/ra_hir/src/debug.rs [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+pub trait HirDebugDatabase {
+    fn debug_crate(&self, krate: Crate, fmt: &mut fmt::Formatter<'_>) -> fmt::Result;
+    fn debug_module(&self, module: Module, fmt: &mut fmt::Formatter<'_>) -> fmt::Result;
+}
+
+impl<DB: HirDebugHelper> HirDebugDatabase for DB {
+    fn debug_crate(&self, krate: Crate, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let mut builder = fmt.debug_tuple("Crate");
+        match self.crate_name(krate.crate_id) {
+            Some(name) => builder.field(&name),
```
Comment (bjorn3) [inline crates/ra_hir/src/debug.rs]: Maybe show crate id in any case. That would prevent confusion with multiple versions of a crate.

Comment (matklad) [reply crates/ra_hir/src/debug.rs]: I prefer to print less by default, as "too verbose logging" is harder to fix in a specific case. It's possible to print the `Crate` itself to see the ID

**Pull request merged**